overlord/snapstate/snapstate.go

```diff
@@ -256,6 +257,48 @@ func optedIntoSnapdSnap(st *state.State) (bool, error) {
 	return experimentalAllowSnapd, nil
 }
 
+// refreshRetain returns refresh.retain value if set, or the default value (different for core and classic).
+// It deals with potentially wrong type due to lax validation.
+func refreshRetain(st *state.State) int {
+	var val interface{}
+	// due to lax validation of refresh.retain on set we might end up having a string representing a number here; handle it gracefully
+	// for backwards compatibility.
+	err := config.NewTransaction(st).Get("core", "refresh.retain", &val)
+	var retain int
+	if err == nil {
+		switch v := val.(type) {
+		// this is the expected value; confusingly, since we pass interface{} to Get(), we get json.Number type; if int reference was passed,
+		// we would get an int instead of json.Number.
+		case json.Number:
+			retain, err = strconv.Atoi(string(v))
```

Comment:
Kind of nitpick/question: were `int64` numbers rejected or allowed before? If `int64` numbers were allowed I think they would just be truncated to `int` before which means this isn't 100% compatible because Atoi will error in that case. We could do something like this to truncate:
```suggestion
			var num int64
			num, err = v.Int64()
			retain = int(num)
```

It's a bit of a nitpick because users probably don't set values that large. However, I could see someone doing that if they wanted to keep every revision around for some reason

Reply:
We constrain retain value to 2..20 range; this is enforced on Set and works also when a string is stored in config.

Reply:
Ah, nevermind then. Thanks!